### PR TITLE
Adding custom queries for Oracle

### PR DESCRIPTION
**Pull request by nenadnoveljic**
### What does this PR do?

Add custom queries to Oracle monitoring.

### Motivation

The feature was requested by a beta customer.

### Describe how to test/QA your changes

Add the following configuration and check that the metrics are appearing in the front end:

```
      - metric_prefix: oracle.custom_query.locks
        query: |
          select nvl(blocking_session,0), username, osuser, sid, serial# as serial, wait_class, seconds_in_wait
          from GV$SESSION
          where rownum<=5
        columns:
          - name: blocking_session
            type: gauge
          - name: username
            type: tag
          - name: osuser
            type: tag
          - name: sid
            type: tag
          - name: serial
            type: tag
          - name: wait_class
            type: tag
          - name: seconds_in_wait
            type: tag
```

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by hestonhoffman — APPROVED** [review]

File: pkg/collector/corechecks/oracle-dbm/custom_queries.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+				if q.Columns[i].Type == "tag" {
+					if v != nil {
+						tags = append(tags, fmt.Sprintf("%s:%s", q.Columns[i].Name, v))
+					}
+				} else if methodFunc, ok := methods[q.Columns[i].Type]; ok {
+					metricRow.name = fmt.Sprintf("%s.%s", metricPrefix, q.Columns[i].Name)
+					if v_str, ok := v.(string); ok {
+						metricRow.value, err = strconv.ParseFloat(v_str, 64)
+						if err != nil {
+							log.Errorf("Metric value %v for metricRow.name %s is not a number", v, metricRow.name)
```
Comment (justiniso) [inline pkg/collector/corechecks/oracle-dbm/custom_queries.go:103]: These are user-facing logs, right? If I was a user and saw this line in the logs, I probably would not know how to respond. This log with no context doesn't tell me much:

`2023-07-05 15:06:53 UTC | CORE | INFO | (pkg/collector/oracle-dbm/custom_queries.go:46 in RunAll) | Metric value octo for metricRow.name cat is not a number`

I would prefer to see additional information and context like this: 

`2023-07-05 15:06:53 UTC | CORE | INFO | (pkg/collector/oracle-dbm/custom_queries.go:46 in RunAll) | Custom query #1 encountered a type error during execution. A number was expected for the column "name" but the query results returned the value "octo" (query was: "SELECT user1, user2, ...")`

**Review by dujuku — COMMENTED** [review]
None of my comments are blocking, but I had some suggestions for improvements.

File: pkg/collector/corechecks/oracle-dbm/custom_queries.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+type metricRow struct {
+	name   string
+	value  float64
+	method Method
+	tags   []string
+}
+
+func (c *Check) CustomQueries() error {
+	if c.dbCustomQueries == nil {
+		db, err := c.Connect()
```
Comment (justiniso) [inline pkg/collector/corechecks/oracle-dbm/custom_queries.go:29]: I don't fully understand why we're opening a new connection here. Can you explain?

Comment (nenadnoveljic) [reply pkg/collector/corechecks/oracle-dbm/custom_queries.go:29]: Custom queries is the only feature that switches to PDBs (all other queries are running against the root container). Switching to PDB and subsequent query execution isn't atomic, so there's no guarantee the both operations would get the same connection from the pool.

File: pkg/collector/corechecks/oracle-dbm/config/config.go [hunk 3]
```diff
@@ -60,6 +60,19 @@ type AgentSQLTrace struct {
 	TracedRuns int  `yaml:"traced_runs"`
 }
 
+type CustomQueryColumns struct {
+	Name string `yaml:"name"`
+	Type string `yaml:"type"`
+}
+
+type CustomQueries struct {
```
[the remark this answers is not in the document]
Comment (nenadnoveljic) [reply pkg/collector/corechecks/oracle-dbm/config/config.go]: I changed it to the singular.

File: pkg/collector/corechecks/oracle-dbm/custom_queries.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+		c.dbCustomQueries = db
+		c.dbCustomQueries.SetMaxOpenConns(1)
+	}
+
+	var metricRows []metricRow
+	sender, err := c.GetSender()
+	if err != nil {
+		return fmt.Errorf("failed to get sender for custom queries %w", err)
+	}
+	methods := map[string]Method{"gauge": sender.Gauge, "count": sender.Count, "rate": sender.Rate}
```
[the remark this answers is not in the document]
Comment (nenadnoveljic) [reply pkg/collector/corechecks/oracle-dbm/custom_queries.go]: I added other methods as well.

File: pkg/collector/corechecks/oracle-dbm/custom_queries.go (hunk 1 above)
Comment (nenadnoveljic) [reply pkg/collector/corechecks/oracle-dbm/custom_queries.go]: I improved the error messages.

File: pkg/collector/corechecks/oracle-dbm/custom_queries.go (hunk 2 above)
Comment (justiniso) [reply pkg/collector/corechecks/oracle-dbm/custom_queries.go:36]: Makes sense! Would appreciate a comment indicating this 🙏

**Review by dujuku — APPROVED** [review]

**Pull request merged**